### PR TITLE
Federation services, complete

### PR DESCRIPTION
**Pull request by gmac**
@tvvignesh – some knowledge on hooking federation directly into stitching here if that's something you're looking to do.

### CODE REVIEW

**Comment by tvvignesh** [comment]
@gmac Thanks. But since I have completely moved away from federation to stitching now, this may no longer be needed for me (I hope stitching works well for my usecase as I go along, but I will learn along the way anyways).

Btw, thanks a ton for this: https://github.com/gmac/schema-stitching-demos/tree/master/hot-schema-reloading 

This was something I had been discussing with @yaacovCR and I was looking for a reference implementation I can use with stitching directives and that definitely helps a lot. I am going with this 🎉

Just have one doubt though. Let's say that the upstream SDL of one of the services is invalid (I don't know when/if this can happen since I can't put my mind around it), but in such cases would the gateway use the valid SDL / invalid SDL

Sorry for the lack of activity/response so far. I was working on some of the issues I was facing with Kubernetes and had not concentrated on this. Getting back on track now.

**Comment by yaacovCR** [comment]
Federation SDL to valid SDL is really cool, and I love how you then construct the merge config automatically. Wondering if we should export this as helper perhaps within a separate library or within graphql-tools.... Should endpoint be final merge config as you have, or just directives that fit stitching directives format...

**Comment by gmac** [comment]
@tvvignesh — I do have ambitions to make a “Testing and Versioning” example, though it’ll be more involved and will have to put some thought into it. Long story short is that you commit schemas into a registry where they are tested together, and then have hot reload simply hit the registry for a precomposed version that increments. You can read about how Apollo does it here: https://www.apollographql.com/docs/federation/managed-federation/overview/. Personally, I find their approach to be a little sloppy in that they consider composition === integrity, which is not always the case in my experience (i.e: you can have a perfectly valid schema that breaks integrity). We still don’t do any kind of hot reloads at my company; we commit schemas into the gateway app’s repo where they have mocked service tests run against the composed schema, then redeploy. It’s slow, but the integrity is very high. We’ve talked recently about putting some effort into designing a system that maintains our integrity while speeding up versioning and deployment. This future example may be a rough prototype of it!

@yaacovCR — converting fed SDL to stitching SDL is interesting, but seems like an extra step in the ultimate goal of SDL to subschema config. Either way, my only hesitance to a formal implementation somewhere is the maintenance... one of the nice things of an example like this is that it shows you pretty directly how to do it without promising that it’ll work for your exact schema which inevitably always contains some bizarre outlier.

**Comment by tvvignesh** [comment]
@gmac @yaacovCR I used https://github.com/gmac/schema-stitching-handbook/tree/master/hot-schema-reloading as the inspiration, made some changes and I can confirm that the schema hot reloading is working great when any of the subservices are changed

I also had a look at the new Video series about local/remote schemas, single record typemerging, array batching made by you and this really is fantastic. Clearly explained (Wish I had this a month back 😅)

Working on Type merging now with stitching-directives. Will update this thread.

**Comment by gmac** [comment]
Awesome!! I’ve also got a demo of hot reloading in the works using a GitHub repo as a schema registry to poll. Should have that demo released in the next week

**Comment by tvvignesh** [comment]
@gmac Great! I use Gitlab though (but since it is working by loading a remote SDL anyways I can work with this for now).

Also, I had a doubt. I wanted to do array batched merging using stitching directives and did refer https://github.com/gmac/schema-stitching-handbook/tree/master/stitching-directives-sdl but did not find an example for that.

So I went through the tests @yaacovCR had written. Is this the one to be used for array batched type merging?

```ts
test('adds key and argsFromKeys functions with argsExpr argument using a fully qualified key', () => {
    const typeDefs = `
      ${stitchingDirectivesTypeDefs}
      scalar _Key
      type Query {
        _user(key: _Key): [User] @merge(argsExpr: "key: [[{ id: $key.id }]]")
      }
      type User @key(selectionSet: "{ id }") {
        id: ID
        name: String
      }
    `;
```

**Comment by tvvignesh** [comment]
I am not sure, but does your example here: https://github.com/gmac/schema-stitching-handbook/blob/master/stitching-directives-sdl/services/products/schema.graphql use array batched type merging? Cause I see `keyField` which I guess is part of array batched merging. Am I right in my assumption?

**Comment by gmac** [comment]
I actually don’t know. I’m not very familiar with the SDL api yet, that’s
all @yaacovCR. I’ve got a draft PR open for distilling knowledge down into
a formal docs page. It has a ways to go. For me, I still don’t understand
how _Key works.

**Comment by tvvignesh** [comment]
@yaacovCR @gmac Got type merging working finally with stitching-directives 🎉🎉 

Currently just went with batched merging. Will probably look at array batched merging later after I get some more clarity on it.

Thanks for all your help. It actually turned out to be quite simple 😀

**Comment by tvvignesh** [comment]
Btw `@merge` has too many keys though: `directive @merge(argsExpr: String, keyArg: String, keyField: String, key: [String!], additionalArgs: String) on FIELD_DEFINITION`

`keyArg`,`keyField`,`key` - All of which are not explanatory to be honest. But anyways, I don't want to bother you guys again with this. So, I will leave it at this.

**Comment by gmac** [comment]
@tvvignesh – I just added notes to the in-process chapter about Registry and Versioning... https://github.com/gmac/schema-stitching-handbook/compare/versioning?expand=1

While the narration isn't done yet and some of the code will get cleaned up, it all works and the instructions are complete enough that they should walk you through it. That uses a GitHub repo as a schema registry to manage versioned releases that get hot-reloaded into the production gateway when commits are made.

**Comment by yaacovCR** [comment]
There is no difference between batch merging and regular merging, it is introspected automatically from the return type of the resolver that you attach the merge directive to.. keyArg is when you want to specify which argument that you should send the key to for that resolver... keyField is for when you want to use only a simple key rather than an object and so you specify which field from the selection set will be that simple key, key is when you need to customize how to map the selection set to the argument via a custom key....

Basically if you have got it working with without batching, you should just be able to change the return type to a list and everything should work automatically

**Comment by tvvignesh** [comment]
@yaacovCR Thanks for the clarification (Still trying to digest this part alone). Since the JS implementation uses `argsFromKeys`, `keys` for batch merging and `args` for regular merging, I was looking to understand if I have to change the directives as well similar to that. But from what you are saying, I guess it infers it automatically. 

Currently something like `@merge(keyField: "id")` is working well for me since all my types have an ID field which is what I would be using to stitch together multiple types from multiple services. So, I guess I need not use the rest like `keyArg` or `key`

**Comment by tvvignesh** [comment]
After doing this, I feel like I have more control with the schema on my end unlike federation where a lot of things were abstracted away from me. While that was simple, I feel like this is powerful to work with as we scale and the modular nature of stitching with `graphql-tools` would help in just using the parts as needed without any bloat. Liking it ❤ And with this handbook now in place along with the video series by @gmac new devs should feel right at home 👍

I would say that if the default naming of the arguments is reviewed again, I don't think there is any other problem with stitching.

Also, maybe the gateway can also become a lib in the future like `@graphql-tools/gateway` with built in support for stitching-directives, hot reloading, and so on which people can use if needed or build their own putting everything together.

And the registry can become `@graphql-tools/registry` with multiple backends like `@graphql-tools/backend-github` , `@graphql-tools/backend-localstorage` and so on in the future.

A lot of potential to expand this especially keeping this handbook as a reference implementation of everything.

**Comment by gmac** [comment]
@tvvignesh - awesome, glad that versioning demo was insightful. All the instructions worked okay? Also stay tuned, next up after that one will be testing schemas, so that you can be running CI on your registry using your own validations that you can be as stringent or as lax as you want to be with... versus federation where it just checks if your schemas compose together and green-lights the release assuming they do.
